lms/extractors/ziparchive.py

```diff
@@ -30,35 +30,65 @@ def can_extract(self) -> bool:
         return self.is_zipfile
 
     @staticmethod
-    def _extract(archive: ZipFile, filename: str) -> File:
+    def _extract(archive: ZipFile, filename: str, dirname: str = '') -> File:
         with archive.open(filename) as current_file:
             log.debug(f'Extracting from archive: {filename}')
             code = current_file.read()
         decoded = code.decode('utf-8', errors='replace').replace('\x00', '')
+        filename = filename[len(dirname):]
         return File(path=f'/{filename}', code=decoded)
 
-    def get_exercise(self, file: ZipFile) -> Tuple[int, List[File]]:
+    def get_files(
+        self, archive: ZipFile, filenames: List[Text], dirname: str = '',
+    ) -> Iterator[File]:
+        unwanted_files = self.get_unwanted_files(filenames)
+        yield from(
+            self._extract(archive, filename, dirname)
+            for filename in filenames
+            if (
+                filename.startswith(dirname)
+                and filename not in unwanted_files
+                and filename != dirname
+            )
+        )
+
+    def get_exercises_by_dirs(
+        self, archive: ZipFile, filenames: List[Text],
+    ) -> Iterator[Tuple[int, List[File]]]:
+        for dirname in filenames:
+            parent_name, _ = os.path.split(dirname)
+            exercise_id, _ = self._clean(parent_name)
+            if (
+                exercise_id
+                and len(dirname.strip(os.path.sep).split(os.path.sep)) == 1
+            ):
+                # Checking if the dirname is the first dir in the zipfile
+                # and the first dir is the exercise id.
+                files = list(self.get_files(archive, filenames, dirname))
+
+                yield exercise_id, files
+
+    def get_exercise(self, file: ZipFile) -> Iterator[Tuple[int, List[File]]]:
         assert self.filename is not None
         exercise_id, _ = self._clean(self.filename.rpartition('.')[0])
-        if not exercise_id:
-            raise BadUploadFile('Invalid zip name', self.filename)
 
         with file as archive:
-            namelist = archive.namelist()
-            unwanted_files = self.get_unwanted_files(namelist)
-
-            files = [
-                self._extract(archive, filename)
-                for filename in namelist
-                if filename not in unwanted_files
-            ]
-
-        return exercise_id, files
+            filenames = archive.namelist()
+            if not exercise_id:
```

Comment:
We not just `yield from self.get_exercises_by_dirs(archive, filenames)`?

(Also: flip the condition to be `if exercise_id`)